apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+                // For acks and nacks, ackActionToRetry must be non-null and specify the right queue (_ackQueue or _nackQueue).
+                // For extends, extendActionToRetry must be non-null and specify the right queue,i.e, _extendQueue.
+                void RetryTemporaryFailures(
+                    IEnumerable<string> ackOrNackIds,
+                    Action<IEnumerable<string>> ackActionToRetry,
+                    IEnumerable<TimedId> extendIds,
+                    Action<IEnumerable<TimedId>> extendActionToRetry = default,
+                    int timeoutInSeconds = 600)
+                {
+                    // TemporaryFailures may be empty but will never be null, so null check is not needed.
+                    // By design, only items in temporaryFailures that exist in _retyrableIds will be retried.
+                    // If id is not in _retryableIds dictionary, it is either already succeeded or not retryable.
+                    // Failure may be non-retryable due to exceeded time duration or that failure may have become permanent.
+
+                    bool hasAcksOrNacks = ackOrNackIds.Any();
+                    bool hasExtends = extendIds.Any();
+                    if (!hasAcksOrNacks && !hasExtends)
+                    {
+                        return;
+                    }
+
+                    var idsToRetry = hasAcksOrNacks ? ackOrNackIds : extendIds.Select(j => j.Id);
+                    var retryableErrorIds = new ConcurrentBag<string>();
+                    // Not all temporaryFailures may need to be retried.
+                    // See if the id exists in the _retryableIds dictionary and only then retry.
+                    foreach (var id in idsToRetry.Where(j => _retryableIds.ContainsKey(j)))
+                    {
+                        retryableErrorIds.Add(id);
+                    }
+
+                    if (retryableErrorIds.Count == 0)
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    // All ids in retryableErrorIds will have the same backoff.
+                    // Get the retryInfo for the first retryableErrorId from the _retryableIds dictionary.
+                    if (!_retryableIds.TryGetValue(retryableErrorIds.First(), out var retryInfo))
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
+                    // We should retry only for specified timeout.
+                    // If current UTC time is greater than first UTC time of failure + timeout, we should not retry.
+                    if (DateTime.UtcNow > retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(timeoutInSeconds) - backoff)
+                    {
+                        // We are past the retry timeout.
```

Comment:
So presumably these are now permanent failures... but how do they become permanent?

Reply:
If no successful ack or modack is made for a message id for 10 mins, server will treat that ack id as expired. That's because the max ack deadline can be 10 mins. After that, an ack/modack will fail with a permanent error since the previous ack is now invalid. So, the Pub/Sub team doesn't want us to retry after 10 minutes.